botshot/core/scheduler.py

```diff
@@ -138,27 +112,143 @@ def _schedule_wrapper(self, user_spec: dict, action, task_id=None):
+            return dt
+        if date is None:  # weekday
+            day_delta = weekday - now.weekday
+            dt = now.replace(hour=hour, minute=minute, second=second) + timedelta(days=day_delta)
+            if dt <= now:
+                dt += timedelta(days=7)
+            return dt
+        if month is None: # date
+            dt = datetime(year=now.year, month=now.month, day=date, hour=hour, minute=minute, second=second)
+            if dt <= now:
+                month, year = now.month + 1, now.year if now.month < 12 else 1, now.year + 1
+                dt = datetime(year=year, month=month, day=date, hour=hour, minute=minute, second=second)
+            return dt
+        dt = datetime(year=now.year, month=month, day=date, hour=hour, minute=minute, second=second)
+        if dt <= now:
+            dt.replace(year=now.year+1)
+        return dt
+
+    def _validate_timespec(self, timespec: dict):
         try:
-            crontab(**kwargs)
+            if timespec.get('day_of_month') is not None and timespec.get('day_of_week') is not None:
+                raise Exception("Timespec must contain at most one of: day_of_month, day_of_week")
+            
+            keys = ['minute', 'hour', 'day_of_month', 'month_of_year']
+            ranges = {
+                "second": (0, 59),
+                "minute": (0, 59),
+                "hour":   (0, 23),
+                "day_of_month":  (1, 31),
+                "day_of_week":   (0, 6),
+                "month_of_year": (1, 12)
+            }
+
+            was_present = True
+            last_key = "second"
+            key_cnt = 1 if timespec.get('second') is not None else 0
+
+            for key in keys:
+                # every time interval requires the previous one to be present
+                # for example, to repeat a task every hour we also need to know the minute
+                is_present = timespec.get(key) is not None
+                if key == 'day_of_month':
+                    is_present |= timespec.get('day_of_week') is not None
+                if is_present and not was_present:
+                    raise Exception("When using %s in timespec, %s also needs to be present" % (key, last_key))
+                
+                was_present = is_present
+                last_key = key
+                if is_present: key_cnt += 1
+            
+            for key, range in ranges.items():
+                value = timespec.get(key)
+                if value is not None:
+                    if value < range[0] or value > range[1]:
+                        raise Exception("%s must be in range (%d, %d)" % (key, range[0], range[1]))
+
+            if key_cnt <= 0:
+                raise Exception("Timespec must contain at least one time key")
+
         except Exception as e:
             raise Exception("Invalid schedule specification") from e
 
-    def _validate_user(self, user):
-        pass  # TODO
+    def _validate_user_spec(self, user_spec: dict):
```

Comment:
If the conversion to JSON is generic, the user should be able to implement their own `ConversationFilter` in their own project code...

Reply:
```python
from abc import ABC, abstractmethod

class ConversationFilter(ABC):
	
	@abstractmethod
	def get_ids():
		pass


class ListConversationFilter(ConversationFilter):
	def __init__(ids):
		super().__init__()
		self.ids = ids

	def get_ids():
		return ids


class AllConversationFilter(ConversationFilter):

	def get_ids():
		return [c.id for c in ChatConversation.objects.all()]

```

Reply:
core/persistence.py

Not tested :P 

```python
from django.utils.module_loading import import_string

class DictSerializable:
    """
    Marker class used only to specify that our class can be serialized using its __dict__ field
    Inheriting from this class is enforced instead of trying to serialize any object to avoid unexpected errors when serializing objects that should not be serialized
    """
    pass

def fullname(o):
    module = o.__class__.__module__
    if module is None or module == str.__class__.__module__:
        return o.__class__.__name__  # Avoid reporting __builtin__
    return module + '.' + o.__class__.__name__

def json_deserialize(obj):
    #print('Deserializing:', obj)
    obj_type = obj.get('__type__')
    if obj_type == 'datetime':
        return dateutil.parser.parse(obj.get('value'))
    elif obj_type == 'entity':
        bytearr = str.encode(obj.get("__data__"))
        return pickle.loads(b64decode(bytearr))
    elif obj_type:
        return import_string(obj_type)(**obj.get("__dict__"))
    return obj

def json_serialize(obj):
    from datetime import datetime
    # from botshot.core.entities import Entity
    if isinstance(obj, datetime):
        return {'__type__':'datetime', 'value': obj.isoformat()}
    elif isinstance(obj, EntityValue):
        data = b64encode(pickle.dumps(obj))
        return {"__data__": data.decode('utf8'), '__type__': 'entity'}
    elif isinstance(obj, DictSerializable):
        return {'__dict__': obj.__dict__, '__type__': fullname(obj)}
    return obj

```

Reply:
```
    def add_schedule(self, action, conversations, at: datetime, description=None) -> str:
        if not isinstance(conversations, ConversationFilter):
            conversations = ListConversationFilter(self._get_conversation_ids(conversations))
        ...

    def _get_conversation_ids(conversations):
        if isinstance(conversations, int):
            return [conversations]
        elif isinstance(conversations, ChatConversation):
            return [conversation.id]
        elif isinstance(conversations, list):
            ids = []
            for c in conversations:
                ids += self._get_conversation_ids(c)
            return ids
        raise ValueError('Invalid conversations filter, pass instance of ConversationFilter or an instance (or list) of ChatConversations or conversation IDs.')
```